geopandas/sindex.py

```diff
@@ -70,6 +91,150 @@ def __init__(self, geometry):
+            """Compatibility layer for pygeos.query.
+
+            This is not a vectorized function, if speed is important,
+            please use PyGEOS.
+
+            Parameters
+            ----------
+            geometry : shapely geometry
+                A single shapely geometry to query against the spatial index.
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
+                If predicate is provided, the input geometry is tested using the predicate function against each
+                item in the tree whose extent intersects the envelope of the input geometry:
+                predicate(input_geometry, tree_geometry).  If possible, prepared geometries are used to help
+                speed up the predicate operation.
+            sort : bool, default False
+                If True, the results will be sorted in ascending order. If False, results are
+                often sorted but there is no guarantee.
+
+            Returns
+            -------
+            matches : ndarray of shape (n_results, )
+                Integer indices for matching geometries from the spatial index.
+            """  # noqa: E501
+
+            # handle invalid predicates
+            if predicate not in self.valid_query_predicates:
+                raise ValueError(
+                    "Got `predicate` = `{}`, `predicate` must be one of {}".format(
+                        predicate, self.valid_query_predicates
+                    )
+                )
+
+            # handle empty / invalid geometries
+            if geometry is None:
+                # return an empty integer array, similar to pygeys.STRtree.query.
+                return np.array([], dtype=np.intp)
+
+            if not isinstance(geometry, BaseGeometry):
+                raise TypeError(
+                    "Got `geometry` of type `{}`, `geometry` must be ".format(
+                        type(geometry)
+                    )
+                    + "a shapely geometry."
+                )
+
+            if geometry.is_empty:
+                return np.array([], dtype=np.intp)
+
+            # query tree
+            bounds = geometry.bounds  # rtree operates on bounds
+            tree_idx = list(self.intersection(bounds, objects=False))
+
+            if not tree_idx:
+                return np.array([], dtype=np.intp)
+
+            # Check predicate
+            # When possible, we want to use cached prepared geometries
+            # from the tree to compare tree_geom.predicate(input_geom)
+            # If this is not possible, try prep the input geometry
+            # and compare that against results from the tree.
+            # Since only certain predicates support prepared geometries,
+            # the fallback is to check non-prepared geometries.
+            if predicate in ("intersects", "within"):
+                # For these two predicates, we compare tree_geom.predicate(input_geom)
```

Comment:
For "within" I understand now why we are flipping this around. But why do this for "intersects"? I think it will be more efficient to prepare the input geometry, since this is called multiple times

Reply:
Well we can cache the tree geometry, but not the input geometries. So I guess it depends on the specific dataset. I did it this way so that we are caching whenever possible, but I'm open to doing it the other way around.

Reply:
Assuming a single call to query_bulk, each geometry will be passed once to query, so I think it is OK they are not cached. 

(we currently also don't do this, and caching prepared geoms in general is a separate topic we can discuss with pygeos, I would say. We shouldn't put too much effort in optimizing the current shapely code, at least not beyond what we have now. As it will become obsolete soon)

Reply:
(BTW, this is also how it is done in pygeos now (prepare the input geometry, not the tree geometry)

Reply:
Okay, no problem, I can move intersects to the other loop so that the input geometry is prepared. Do we still want to cache the prepared tree geometry to "within", or should I get rid of that too?

Reply:
Keeping it for "within" might be fine. We do switch the order in `sjoin`, so in practice are using prepared geoms there.

Reply:
Right, I got the idea from sjoin. Once we put `bulk_query` into sjoin, my idea would be to remove prepared geometries from sjoin (they're not compatible with pygeos).